### ISSUE TITLE
Support range queries for JMT/Storage data

### ISSUE THREAD

**Issue by zbuc** [penumbra-zone/penumbra#1532]
In #1491 we implemented the base of the new `Storage` implementation. Part of this included storing every JMT key preimage in a `jmt_keys` RocksDB column family.

Now we would like to be able to deal with retrieving ranged data, for example retrieving all unfinished governance proposal keys matching a prefix like: `governance/unfinished_proposals/*`

To implement this, we will want to use RocksDB's `PrefixRange` to specify the prefix: https://docs.rs/rocksdb/latest/rocksdb/struct.ReadOptions.html#method.set_iterate_range

We should first extend the `StateRead` trait with something similar to the following:

```
    fn get_prefix<D, P>(&self, prefix: &str) -> Pin<Box<dyn Stream<Item = Result<(D), Self::Error>> + Send + '_>>
    where
        D: Protobuf<P>,
        P: Message + Default,
        P: From<D>,
        D: TryFrom<P> + Clone + Debug,
        <D as TryFrom<P>>::Error: Into<anyhow::Error>;
```

And then provide implementations for `Snapshot` (that retrieves the data directly from the snapshot) and for `State` (that handles interleaving unsaved data).

**Comment by hdevalence** [comment]
We're never writing to a prefix, only reading, so I think we could shorten the method name to just `prefix`. We could also split the implementation along the same lines as `get`, so that we'd have `prefix_raw`, `prefix_proto`, and `prefix` (which returns domain types).

Then the `prefix` and `prefix_proto` methods could be provided by the trait implementation (layering proto or domain decoding on top of calls to `prefix_raw`, and only the `prefix_raw` method would need to be defined.

Inside of the implementation, we'll need to have some logic that interleaves results from the underlying rocksdb storage, and from the internal `BTreeMap` caches. We might end up with some duplication, but perhaps we could make an internal helper function that interleaves results from (1) rocksdb (2) a `&BTreeMap` (3) an `Option<&BTreeMap>`, so that we can use it for both reads from the `State` and from a `StateTransaction`.

I don't think we need an implementation of the `StateRead` trait for `Snapshot`, because we'll never want to be working with a `Snapshot` directly, only one wrapped up inside of a `State`.

**Comment by plaidfinch** [comment]
Preserving a comment from discord here:

I actually think we probably want to provide an `Iterator`, not a `Stream`, because the rocks interface is synchronous, and the modification i made to the JMT means that you can pass in an `Iterator<(KeyHash, Option<OwnedValue>)>` to `put_value_sets`, which means we could compose the range queries of the public state with the JMT update creation by iterating over a range and (flat)mapping/filtering as required to form the updates desired to that set of keys. This would be efficient, and also would prevent the "async inside sync" problem.

**Comment by zbuc** [comment]
@plaidfinch I think your comment is out of date now that the rest of the `get` interfaces are asyncified?

**Comment by plaidfinch** [comment]
@zbuc That's correct; I think we don't care about this anymore. Even though it could reduce memory consumption a bit, it would add a significant implementation and interface complexity. Let's not do it.

**Comment by zbuc** [comment]
Closed by #1546
